File: Hash Tables/My Own Hash Table/moht.ts
// class HashTable {
//     private data: Array<Array<[string, any] | undefined>>;
  
//     constructor(size: number) {
//       this.data = new Array(size);
//     }
  
//     private _hash(key: string): number {
//       let hash = 0;
//       for (let i = 0; i < key.length; i++) {
//         hash = (hash + key.charCodeAt(i) * i) % this.data.length;
//       }
//       return hash;
//     }
  
//     set(key: string, value: any): void {
//       const address = this._hash(key);
//       if (!this.data[address]) {
//         this.data[address] = [];
//       }
//       this.data[address].push([key, value]);
//     }
  
//     get(key: string): any | undefined {
//       const address = this._hash(key);
//       const currentBucket = this.data[address];
//       console.log(currentBucket);
//       if (currentBucket) {
//         for (let i = 0; i < currentBucket.length; i++) {
//           if (currentBucket[i][0] === key) {
//             return currentBucket[i][1];
//           }
//         }
//       }
//     }
  
//     keys(): string[] | undefined {
//       if (!this.data.length) {
//         return undefined;
//       }
//       const result: string[] = [];
//       for (let i = 0; i < this.data.length; i++) {
//         if (this.data[i] && this.data[i].length) {
//           if (this.data.length > 1) {
//             for (let j = 0; j < this.data[i].length; j++) {
//               result.push(this.data[i][j][0]);
//             }
//           } else {
//             result.push(this.data[i][0][0]);
//           }
//         }
//       }
//       return result;
//     }
//   }
  
//   const myHashTable = new HashTable(50);
//   myHashTable.set("mangos", 2741);
//   myHashTable.set("bananas", 2137);
//   myHashTable.get("bananas");
  

File: Hash Tables/My Own Hash Table/moht.py
class HashTable:
    def __init__(self, size):
        self.size = size
        self.data = [None] * size

    def _hash(self, key):
        hash_value = 0
        for i in range(len(key)):
            hash_value = (hash_value + ord(key[i]) * i) % self.size
        return hash_value

    def set(self, key, value):
        address = self._hash(key)
        if self.data[address] is None:
            self.data[address] = []
        self.data[address].append([key, value])

    def get(self, key):
        address = self._hash(key)
        current_bucket = self.data[address]
        if current_bucket:
            for item in current_bucket:
                if item[0] == key:
                    return item[1]
        return None

    def keys(self):
        keysArray = []
        for i in range(0, len(self.data)):
            if self.data[i]:
                keysArray = keysArray + [self.data[i][0][0]]
        return keysArray


# Create a HashTable instance
myHashTable = HashTable(50)

# Add some key-value pairs
myHashTable.set("mangos", 2741)
myHashTable.set("bananas", 2137)
myHashTable.set("kiwis", 3322)

# Retrieve a value by key
result = myHashTable.get("bananas")
keys = myHashTable.keys()
print(keys)
